### ISSUE TITLE
.NET Core 3.0 Docker Container Won't Connect to SQL Server

### ISSUE THREAD

**Issue by jeffgbradley2** [dotnet/SqlClient#222]
# .NET Core 3.0 Docker Container Won't Connect to SQL Server

We have upgraded some of our .NET Core 2.2 applications to .NET Core 3.0. After the upgrade, our applications running in Docker containers fail to connect to SQL Server 2016 running on a standalone machine in our domain network. Reviewing the packet trace, we see that the connection hangs at the Pre-Login handshake, and eventually times out. Switching back to .NET Core 2.2 resolves the issue immediately, and our containers can again connect to our databases.

We have obviously verified that there are no DNS issues, since these containers are running in production with .NET Core 2.2. .NET Core 3.0 breaks them as they cannot connect to SQL Server over the network, so we are rolling back to 2.2 for now and terminating the upgrade until we can determine why .NET Core 2.2 containers connect to our SQL Servers but .NET Core 3.0 containers do not.

We have also created new .NET Core 3.0 projects from scratch to test the connection. The result is the same: the containers time out during Pre-Login to SQL Server. Changing the project back to .NET Core 2.2, along with EF Core back to 2.2.6, and the issue is immediately resolved.

As we are using EF Core for data access, we have also tried leaving EF Core at 2.2.6 but upgrading the project to .NET Core 3.0, but the connection still times out.

If we do not add Docker support and run the apps as Console or Web apps (not containerized), then there are no issues with .NET Core 3.0. It is only when we containerize them that .NET Core 3.0 apps fail to connect to SQL Server.

Per suggestion in Stack Overflow, I changed the data access in the test project to Microsoft.Data.SqlClient using hand-written SQL and a DataReader. I then recreated the Dockerfile for .net core 3.0 and verified that the correct container image was referenced. Same behavior: hung at PreLogin handshake with SQL Server. Switched to .net core 2.2 and was able to connect and read data from SQL Server.

**Comment by StijnStaelens** [comment]
@vcsjones I tried the extra command, but it still doesn't work. My dockerfile now looks like this
`# Stage 1
FROM mcr.microsoft.com/dotnet/core/sdk:3.0 AS builder
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
WORKDIR src

COPY ./project.csproj .
RUN dotnet restore ./project.csproj
COPY . .
RUN dotnet publish ./project.csproj --configuration Release

# Stage 2
FROM mcr.microsoft.com/dotnet/core/aspnet:3.0
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
COPY --from=builder src/bin/Release/netcoreapp3.0/publish/ app
WORKDIR app

ENTRYPOINT ["dotnet", "project.dll"]`

My code works fine when I run it on windows, in Docker it fails

**Comment by StijnStaelens** [comment]
Using mcr.microsoft.com/dotnet/core/aspnet:3.0-bionic worked.

**Comment by vcsjones** [comment]
@StijnStaelens hm. You're using different base images from @jeffgbradley2, I'll examine those. 

What version of SQL server are you connecting to?

**Comment by StijnStaelens** [comment]
@vcsjones,
We tried with 2 servers. 1 is 'Microsoft SQL Server Web 2017 (64-bit)'. The other is a 'Microsoft SQL Server Web 2012 (64-bit)'. Both failed. These are both self-hosted and we try to connect with this connectionstring:
Password=xxxxx;Persist Security Info=True;User ID=xxxxxx;Initial Catalog=xxxxxx;Data Source=1xx.xxx.xxx.xxx;

**Comment by vcsjones** [comment]
@StijnStaelens can you please try:

```
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /usr/lib/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
```

**Comment by jeffgbradley2** [comment]
We have confirmed that switching to Bionic fixes the issue.

**Comment by jeffgbradley2** [comment]
Switching to bionic fixed our issue for asp.net core, but not for console containers. For console containers, after switching to bionic we get the following:
```
It was not possible to find any compatible framework version
The specified framework 'Microsoft.AspNetCore.App', version '3.0.0' was not found.
  - No frameworks were found.

You can resolve the problem by installing the specified framework and/or SDK.
```

**Comment by MichaelSimons** [comment]
@jeffgbradley2, can you make sure you have the latest version of the images pulled locally?  If you have a previous version cached, you must explicitly pull the images to get any updates.  

Another way to validate would be to run `docker run -it --rm mcr.microsoft.com/dotnet/core/runtime:3.0-bionic dotnet --list-runtimes`

**Comment by jeffgbradley2** [comment]
This is happening locally and in our CICD pipeline.

Running that command is what I and our Devops group expect to see.

**Comment by MichaelSimons** [comment]
@jeffgbradley2 - I'm confused.  You said:

> For console containers, after switching to bionic we get the following:

> It was not possible to find any compatible framework version
> The specified framework 'Microsoft.AspNetCore.App', version '3.0.0' was not found.
>   - No frameworks were found.
> 
> You can resolve the problem by installing the specified framework and/or SDK.

What do you mean by console container?  Are you saying it is a containerized console app?  Why is your console app depending on aspnet?

**Comment by jeffgbradley2** [comment]
Great question. Looks like that framework is added by default. When creating a .Net Core Console Application. Here is screenshot from brand new console app.

![image](https://user-images.githubusercontent.com/51322592/65795953-7727e700-e139-11e9-949d-ee4af1f21f15.png)

**Comment by jeffgbradley2** [comment]
Our Devops has ruled this too fragile for production. Thanks anyway.

**Comment by MichaelSimons** [comment]
@jeffgbradley2 , I'm not following.  Your screen shot shows Microsoft.NETCore.app which is what I expect for a console app.  The error you included was for the Microsoft.AspNetCore.App.

**Comment by jeffgbradley2** [comment]
That is true. It is a mystery. We are giving up on 3.0 for now, so no need to spend anymore time. Thank you anyway.

**Comment by aidanjryan** [comment]
This looks to me like there are two different issues here. The original report is about a hang. A different person (@StijnStaelens ) reported a pre-login handshake error.

I think it's doubtful that TLS implementation would be responsible for the hang, although I guess it's possible. The hang is probably the same as #22, #165, or #201

We're experiencing this hang and it's stopping us from deploying .NET Core 3 across our infra.

**Comment by aidanjryan** [comment]
@jeffgbradley2 your CI troubleshooting sure is making a lot of noise in this issue, and an error about "not possible to find any compatible framework version" sure isn't related to the issue re: connecting to SQL server.

**Comment by StijnStaelens** [comment]
> 
> 
> @StijnStaelens can you please try:
> 
> ```
> RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
> RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
> RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /usr/lib/ssl/openssl.cnf
> RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
> ```

@vcsjones , I now used the standard image mcr.microsoft.com/dotnet/core/aspnet:3.0 (that didn't work before) and added those commands. Now it seems to work.

**Comment by MichaelSimons** [comment]
Should this issue have been closed or should it be left open?  The end user experience doesn't feel optimal to me.  This is going to affect a lot of users, how are they going to know how to address the problem?  Downgrading the `seclevel` feels like a patch.  What changes need to be made to work with `seclevel2`?

**Comment by vcsjones** [comment]
>What changes need to be made to work with seclevel2?

This is ultimately a problem of the SQL server that is being connected to. The snippet I posted does two things: it lowers the required TLS version, and adjusts the SECLEVEL down to one.

Changes the minimum TLS version from 1.2 to 1.0. If this fixed the problem, then the SQL server needs to be patched and configured to use TLSv1.2 as described in [this support article](https://support.microsoft.com/en-us/help/3135244/tls-1-2-support-for-microsoft-sql-server). I'm a bit out of my comfort zone with SQL Server + TLS.

If the SECLEVEL change was the source of the fix, then then this could be a few things, such as RC4 being used, or DHE-1024 getting used.

@StijnStaelens If you could help answer one more question, that'd be great. Given these two fixes:

1:

```
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /usr/lib/ssl/openssl.cnf
```

and

2.

```
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
```

Do either of these resolve the issue without the other? That is, can you use 1 without 2 and it still works? Can you use 2 without 1 and does it work? Or does it only work with both 1 and 2?

**Comment by StijnStaelens** [comment]
@vcsjones 

RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf

These commands alone worked.

**Comment by vcsjones** [comment]
@StijnStaelens thanks you! Based on your experiment, it seems that your SQL instances do not support, or have enabled, TLSv1.2.

You can continue to change the minimum version of TLS to v1.0, which is what those commands are doing, or you can [enable][1] TLSv1.2 on your SQL server instances, then the configuration change in the containers will not be necessary. As far as I am aware, there are two steps for TLSv1.2 needed in the SQL server. Install any necessary patches, and enable it in configuration, both of which are covered in the support article.

[1]: https://support.microsoft.com/en-us/help/3135244/tls-1-2-support-for-microsoft-sql-server

**Comment by john-dieter** [comment]
Changing your docker file to this fixed the issue for me:
FROM mcr.microsoft.com/dotnet/core/runtime:3.1-bionic AS base
WORKDIR /app

FROM mcr.microsoft.com/dotnet/core/sdk:3.1-bionic AS build

**Comment by bmart102** [comment]
I had similar issue connecting to remote sql server from linux .net core web app.  I was able to make it work by adding "Connect Timeout=60" to the connection string.

**Comment by jvargasvillegas** [comment]
I can also confirm that using the following image fix the issue:
FROM mcr.microsoft.com/dotnet/core/runtime:3.0-bionic

**Comment by jordankasper11** [comment]
I spent the better part of three days fighting with this same issue before finding this ticket. Switching my Dockerfile to use the 3.1-bionic images instead of 3.1-buster and 3.1-buster-slim allowed me to connect to external SQL Server databases using SqlClient; previous attempts would hang indefinitely.

Can anyone offer any insight into what the underlying problem is? Is there an issue with the buster variants or is there some network issue that Ubuntu-based images are not sensitive to? Before switching the image, none of the TLS-related commands described above made any difference.

**Comment by senvardarsemih** [comment]
> Using mcr.microsoft.com/dotnet/core/aspnet:3.0-bionic worked.

You sir, you made our day ! 3.1-bionic images works for us .

**Comment by reggieboyYEAH** [comment]
I still have the connection problem even with
`mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic`

and this command
```
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
```

I am using a the standard aspnetcore3.1 image from the sample
https://github.com/dotnet/dotnet-docker/blob/master/samples/aspnetapp/Dockerfile

Does anyone know what is the issue?

**Comment by marcelbeeker** [comment]
@reggieboyYEAH I'm have the same issue. My Kubernetes cluster runs on Ubuntu 18.04. My SQL Server is 2016. I try to connect with an ASP.NET Core 3.1 api to this server. 
My docker file is this:
#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.

FROM mcr.microsoft.com/dotnet/core/aspnet:3.1 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build
WORKDIR /src
COPY *.csproj ./
RUN dotnet restore
COPY . .
#WORKDIR "/src/MyFirstDockerApi"
RUN dotnet build "MyFirstDockerApi.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "MyFirstDockerApi.csproj" -c Release -o /app/publish

FROM base AS final
RUN sed -i '/^ssl_conf = ssl_sect$/s/^/#/' /etc/ssl/openssl.cnf
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "MyFirstDockerApi.dll"]

My connectionstring looks like this: Server=<someip>,1433;Database=<datebasename>;User=<username>;Password=<password>;MultipleActiveResultSets=True

When I run this application on my Docker Desktop cluster, I can sucessfully connect. But when I deployed run this image at my Ubuntu cluster, the connection fails. 

The error message I receive is: 
A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 40 - Could not open a connection to SQL Server) Inner exception [A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 40 - Could not open a connection to SQL Server)]

The firewall is diabled. It seems my Kubernetes blocks outgoing traffic but I can't figure out what the problem is.

**Comment by albeserra** [comment]
> @reggieboyYEAH I'm have the same issue. My Kubernetes cluster runs on Ubuntu 18.04. My SQL Server is 2016 running on a standalone machine. I try to connect with an ASP.NET Core 3.1 api to this sql server.
> My docker file is this:
> #See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.
> 
> FROM mcr.microsoft.com/dotnet/core/aspnet:3.1 AS base
> WORKDIR /app
> EXPOSE 80
> EXPOSE 443
> 
> FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build
> WORKDIR /src
> COPY *.csproj ./
> RUN dotnet restore
> COPY . .
> #WORKDIR "/src/MyFirstDockerApi"
> RUN dotnet build "MyFirstDockerApi.csproj" -c Release -o /app/build
> 
> FROM build AS publish
> RUN dotnet publish "MyFirstDockerApi.csproj" -c Release -o /app/publish
> 
> FROM base AS final
> RUN sed -i '/^ssl_conf = ssl_sect$/s/^/#/' /etc/ssl/openssl.cnf
> WORKDIR /app
> COPY --from=publish /app/publish .
> ENTRYPOINT ["dotnet", "MyFirstDockerApi.dll"]
> 
> My connectionstring looks like this: Server=someip,1433;Database=datebasename;User=username>;Password=password;MultipleActiveResultSets=True
> 
> When I run this application on my Docker Desktop cluster, I can sucessfully connect. But when I deployed run this image at my Ubuntu cluster connecting to the same database server, the connection fails.
> 
> The error message I receive is:
> A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 40 - Could not open a connection to SQL Server) Inner exception [A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 40 - Could not open a connection to SQL Server)]
> 
> The firewall is diabled. It seems my Kubernetes blocks outgoing traffic but I can't figure out what the problem is.

I have a similar issue. Have you solved the problem?

**Comment by marcelbeeker** [comment]
No, not yet

Outlook voor Android downloaden<https://aka.ms/ghei36>

________________________________

**Comment by reggieboyYEAH** [comment]
@albeserra  @marcelbeeker I tested today with the latest image. But the issue is not solved yet.
Have you guys figured it out?

**Comment by albeserra** [comment]
> @albeserra @marcelbeeker I tested today with the latest image. But the issue is not solved yet.
> Have you guys figured it out?

not yet... :(

**Comment by aderderian** [comment]
I am seeing this issue with mcr.microsoft.com/dotnet/core/aspnet:3.1-nanoserver-1903 . I can't connect to a remote SQL server. In my case it is Azure. Any updates on this issue? I want to run a windows container.

**Comment by marcelbeeker** [comment]
My problem is solved. I have added an egress rule in my Calico config. Now I can connect to both the internet as SQL Server without any problem.

Outlook voor Android downloaden<https://aka.ms/ghei36>

________________________________

**Comment by aderderian** [comment]
My issue seems to be related to the GlobalProtect gateway as well. It only works when I am on my VPN. Somehow this adapter is conflicting with DNS even when I disable it or change the interface metric order.

**Comment by canertosuner** [comment]
First I tried the set these two commands in my dockerfile


``
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
``
and this works for a while in production env. It didn’t throw any exception for 5-6 hours but after a while it started throwing the same exception.

Than I tried the other suggested solution which is using the **bionic** image.

``
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic
``
and this worked for me and it’s been thrown any exception yet.

**Comment by canertosuner** [comment]
Any update about it ?
I tried using 3.1-bionic image but I got the same exception when there is huge number of request (like 3000 per min.)

**A connection was successfully established with the server, but then an error occurred during the pre-login handshake.**

**Comment by NorDroN** [comment]
Hi!
I have also created new .NET Core 3.1 projects to test the connection to my DB (Microsoft SQL Server 2008 R2 SP1). The result: the containers time out during Pre-Login to SQL Server. I changed the project to .NET Core 2.2 with EF Core 2.2.6, but it didn't help me.

**Comment by hermanho** [comment]
It works without modification on
mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic
but does not work on
mcr.microsoft.com/dotnet/core/aspnet:3.1-focal

**Comment by canertosuner** [comment]
> It works without modification on
> 
> mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic
> 
> but does not work on
> 
> mcr.microsoft.com/dotnet/core/aspnet:3.1-focal

Make a load test like just I did. 200-300 request per second and you will see that it will start throwing handshake exception.

**Comment by aderderian** [comment]
I suggest people look to see if they have a VPN connection or another connection that could potentially be messing with this. I had a GlobalProtect VPN that must have had some routing rules in place even when disabled. I have to have it on for it to work now, despite changing the order of the connection. I didn't have time to solve it further, this was the fix for now.

**Comment by NorDroN** [comment]
It works correct without docker.
It doesn't work in mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic too.

**Comment by hermanho** [comment]
> > It works without modification on
> > mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic
> > but does not work on
> > mcr.microsoft.com/dotnet/core/aspnet:3.1-focal
> 
> Make a load test just like I did. 200-300 request per second and you will see that it will start throwing handshake exception. But if your app does not consume such number of request most probably you will not see any exception.

I tried with jmeter (number of threads is 300). It still works without exception. My app is a form submitting app which insert the form data to sql through ef core. It is running on aks with 1 pod and my sql server is "Microsoft SQL Server 2012 (SP3)". 

I tried to re-build and deploy the image with 3.1-focal and it is stopped working in the first round.

**Comment by niennguyen** [comment]
Use mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic instead of
 mcr.microsoft.com/dotnet/core/aspnet:3.1-buster
It works for me

**Comment by felp95** [comment]
use this image FROM mcr.microsoft.com/dotnet/core/runtime:3.1-bionic AS base 
works for me too thx

**Comment by leili-ncut** [comment]
**FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic**
It works for me  !

**Comment by Leonardo-Ferreira** [comment]
Im having the same issue. Im using the `mcr.microsoft.com/dotnet/core/aspnet:3.1` image and trying to connect to a 2012 SQL Server. Problem started after upgrading from .net core 2.1 to 3.1. On my local machine (windows) I cannot reproduce the issue!

**Comment by felp95** [comment]
It Works for me thanks

**Comment by Atilyus** [comment]
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic
It works for me too

**Comment by Saibamen** [comment]
For `mcr.microsoft.com/dotnet/aspnet:5.0-buster-slim` image I needed to add
```
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /etc/ssl/openssl.cnf
RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/g' /usr/lib/ssl/openssl.cnf
RUN sed -i 's/MinProtocol = TLSv1.2/MinProtocol = TLSv1/g' /usr/lib/ssl/openssl.cnf
```

**Comment by cadaov** [comment]
> We have confirmed that switching to Bionic fixes the issue.

Sadly so far there's no official image for Net 5.
But with this https://github.com/dotnet/SqlClient/issues/222#issuecomment-535941637 it worked

**Comment by cheenamalhotra** [comment]
> Sadly so far there's no official image for Net 5.

Below official images are available for .NET 5:
- mcr.microsoft.com/dotnet/sdk:5.0
- mcr.microsoft.com/dotnet/aspnet:5.0

ref: [Official .NET Docker images](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/net-core-net-framework-containers/official-net-docker-images)

**Comment by Saibamen** [comment]
@cheenamalhotra There are no docker images with .NET 5 and Bionic...

**Comment by cheenamalhotra** [comment]
It is true .NET SDK 5+ are not supporting bionic anymore, but focal images have been available for a while now and with all necessary server and ssl certificate updates, clients should be able to connect via focal images [[Troubleshooting Guide](https://docs.microsoft.com/en-us/sql/connect/ado-net/sqlclient-troubleshooting-guide?view=sql-server-ver15#possible-reasons-and-solutions)]. The most common issue everyone faced with .NET 5 is the update to [default cipher suite](https://docs.microsoft.com/en-us/dotnet/core/compatibility/cryptography/5.0/default-cipher-suites-for-tls-on-linux), for which solutions are available.

Is there a problem that has not been solved yet for TLS 1.2?

Because, everyone would have to do this update sometime to continue upgrading .NET SDK, so not having a bionic image to work with that allows insecure TLS protocols is not going to be an option to block upgrade in future.

I would highly recommend solving environment issues NOW using .NET Core 3.1 focal images and unblocking your applications to be able to upgrade to newer .NET 5+ SDK.

If there are any issues that the SQL Server supports TLS 1.2 but focal images cannot connect and none of the above solutions can solve it, please open a new issue to get the right attention.

**Comment by thinkOfaNumber** [comment]
For those arriving late to the party, none of these options seemed to work for me (neither openssl.cnf nor the 3.1-buster image).

It turns out my problem was the default instance that was listening on 1433 was not the instance that I wanted. I was getting the pre-login handshake error, but with the weird error code `0 - Success`! (This is not the error you're looking for)...
```Microsoft.Data.SqlClient.SqlException (0x80131904): A connection was successfully established with the server, but then an error occurred during the pre-login handshake. (provider: TCP Provider, error: 0 - Success)```

I had to change the default SQL Server instance to use dynamic ports, restart the instance, then change the named instance I wanted to use port 1433, restart _that_ instance, and then everything was great.

Except that I still couldn't use `docker run -p 127.0.0.1:1433:1433` I had to use `docker run -p 1433:1433` otherwise I got yet another error:
```docker: Error response from daemon: Ports are not available: listen tcp 127.0.0.1:1433: bind: An attempt was made to access a socket in a way forbidden by its access permissions. time="2021-08-25T09:47:24+09:30" level=error msg="error waiting for container: context canceled"```

Hope this helps someone following.

**Comment by rimvn** [comment]
> Any update about it ? I tried using 3.1-bionic image but I got the same exception when there is huge number of request (like 3000 per min.)
> 
> **A connection was successfully established with the server, but then an error occurred during the pre-login handshake.**

I faced the same problem. How did you solve it?
